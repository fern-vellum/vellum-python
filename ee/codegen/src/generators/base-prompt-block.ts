import { python } from "@fern-api/python-ast";
import { ClassInstantiation } from "@fern-api/python-ast/ClassInstantiation";
import { MethodArgument } from "@fern-api/python-ast/MethodArgument";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";
import {
  PromptBlock as PromptBlockType,
  FunctionDefinition as FunctionDefinitionType,
} from "vellum-ai/api";

export declare namespace BasePromptBlock {
  interface Args<T extends PromptBlockType | FunctionDefinitionType> {
    promptBlock: T;
  }
}

export abstract class BasePromptBlock<
  T extends PromptBlockType | FunctionDefinitionType
> extends AstNode {
  private astNode: python.ClassInstantiation;

  public constructor({ promptBlock }: BasePromptBlock.Args<T>) {
    super();
    this.astNode = this.generateAstNode(promptBlock);
  }

  protected abstract generateAstNode(promptBlock: T): ClassInstantiation;

  protected constructCommonClassArguments(promptBlock: T): MethodArgument[] {
    const args: MethodArgument[] = [];

    if (promptBlock.state) {
      args.push(
        new MethodArgument({
          name: "state",
          value: python.TypeInstantiation.str(promptBlock.state),
        })
      );
    }

    const cacheConfigValue = this.extractCacheConfig(promptBlock);
    args.push(
      new MethodArgument({
        name: "cache_config",
        value: cacheConfigValue,
      })
    );

    return args;
  }

  private extractCacheConfig(promptBlock: T): python.TypeInstantiation {
    if (
      promptBlock.cacheConfig !== undefined &&
      promptBlock.cacheConfig !== null
    ) {
      if (promptBlock.cacheConfig.type) {
        return python.TypeInstantiation.str(promptBlock.cacheConfig.type);
      }
    }
    return python.TypeInstantiation.none();
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}
