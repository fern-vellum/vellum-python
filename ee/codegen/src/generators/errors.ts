export type CodegenErrorCode =
  | "PROJECT_SERIALIZATION_ERROR"
  | "NODE_ATTRIBUTE_GENERATION_ERROR"
  | "NODE_NOT_FOUND_ERROR";

export abstract class BaseCodegenError extends Error {
  abstract code: CodegenErrorCode;
}

/**
 * An error that raises when deserializing the request to codegen
 * into a valid Workflow project.
 */
export class ProjectSerializationError extends BaseCodegenError {
  code = "PROJECT_SERIALIZATION_ERROR" as const;
}

/**
 * An error that raises when generating a node attribute fails.
 */
export class NodeAttributeGenerationError extends BaseCodegenError {
  code = "NODE_ATTRIBUTE_GENERATION_ERROR" as const;
}

/**
 * An error that raises when a node is not found.
 */
export class NodeNotFoundError extends BaseCodegenError {
  code = "NODE_NOT_FOUND_ERROR" as const;
}
