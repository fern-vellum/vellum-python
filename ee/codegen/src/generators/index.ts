export { ErrorLogFile } from "./error-log-file";
export { InitFile } from "./init-file";
export { Inputs } from "./inputs";
export { NodeInput } from "./node-inputs";
export { Workflow } from "./workflow";
export { VellumVariable } from "./vellum-variable";
export { VellumValue } from "./vellum-variable-value";
