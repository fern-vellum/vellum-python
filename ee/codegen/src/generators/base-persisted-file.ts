import { mkdir, writeFile } from "fs/promises";
import path, { join } from "path";

import { python } from "@fern-api/python-ast";
import { Comment } from "@fern-api/python-ast/Comment";
import { StarImport } from "@fern-api/python-ast/StarImport";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { INIT_FILE_NAME } from "src/constants";
import { WorkflowContext } from "src/context";

export declare namespace BasePersistedFile {
  interface Args {
    workflowContext: WorkflowContext;
    isInitFile?: boolean;
  }
}

export abstract class BasePersistedFile extends AstNode {
  protected readonly workflowContext: WorkflowContext;
  protected readonly isInitFile: boolean;

  public constructor({ workflowContext, isInitFile }: BasePersistedFile.Args) {
    super();
    this.workflowContext = workflowContext;
    this.isInitFile = isInitFile ?? false;
  }

  protected abstract getModulePath(): string[];

  protected abstract getFileStatements(): AstNode[] | undefined;

  protected getFileImports(): StarImport[] | undefined {
    return undefined;
  }

  protected getComments(): Comment[] | undefined {
    return undefined;
  }

  public write(writer: Writer): void {
    const fileStatements = this.getFileStatements();
    if (!fileStatements) {
      return;
    }

    const file = python.file({
      path: this.getModulePath(),
      isInitFile: this.isInitFile,
      statements: fileStatements,
      imports: this.getFileImports(),
      comments: this.getComments(),
    });

    file.inheritReferences(this);

    file.write(writer);
  }

  public async persist(): Promise<void> {
    const absolutePathToModuleDirectory =
      this.workflowContext.absolutePathToOutputDirectory;

    const modulePath = this.getModulePath();

    let fileName: string;
    let filePath: string[];
    if (this.isInitFile) {
      fileName = INIT_FILE_NAME;
      filePath = modulePath;
    } else {
      fileName = modulePath[modulePath.length - 1] + ".py";
      filePath = modulePath.slice(0, -1);
    }

    const filepath = join(absolutePathToModuleDirectory, ...filePath, fileName);
    await mkdir(path.dirname(filepath), { recursive: true });

    const writer = new Writer();
    this.write(writer);

    let contents: string;
    try {
      contents = await writer.toStringFormatted({ line_width: 120 });
    } catch (error) {
      console.error("Error formatting", fileName, "with error", error);
      contents = writer.toString();
    }

    await writeFile(filepath, contents);
  }
}
