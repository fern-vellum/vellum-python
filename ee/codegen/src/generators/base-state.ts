import { python } from "@fern-api/python-ast";

import { WorkflowContext } from "src/context";

export declare namespace BaseState {
  export interface Args {
    workflowContext: WorkflowContext;
  }
}

export class BaseState extends python.Reference {
  public constructor(args: BaseState.Args) {
    super({
      name: "BaseState",
      modulePath: args.workflowContext.sdkModulePathNames.STATE_MODULE_PATH,
    });
  }
}
