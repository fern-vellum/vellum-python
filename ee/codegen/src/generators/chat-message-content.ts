import { python } from "@fern-api/python-ast";
import { MethodArgument } from "@fern-api/python-ast/MethodArgument";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";
import {
  ChatMessageContentRequest as ChatMessageContentRequestType,
  ChatMessageContent as ChatMessageContentType,
} from "vellum-ai/api";

import { VELLUM_CLIENT_MODULE_PATH } from "src/constants";
import { Json } from "src/generators/json";
import { assertUnreachable } from "src/utils/typing";

export namespace ChatMessageContent {
  export interface Args {
    chatMessageContent: ChatMessageContentRequestType | ChatMessageContentType;
    isRequestType?: boolean;
  }
}

export class ChatMessageContent extends AstNode {
  private chatMessageContent:
    | ChatMessageContentRequestType
    | ChatMessageContentType;
  private isRequestType: boolean;

  public constructor({
    chatMessageContent,
    isRequestType = true,
  }: ChatMessageContent.Args) {
    super();
    this.chatMessageContent = chatMessageContent;
    this.isRequestType = isRequestType;
    this.addReferences();
  }

  private addReferences(): void {
    this.addReference(this.getChatMessageContentRef());

    if (this.chatMessageContent.type === "STRING") {
      this.addReference(this.getStringChatMessageContentRef());
    } else if (this.chatMessageContent.type === "FUNCTION_CALL") {
      this.addReference(this.getFunctionCallChatMessageContentRef());
      this.addReference(this.getFunctionCallChatMessageContentValueRef());
    } else if (this.chatMessageContent.type === "AUDIO") {
      this.addReference(this.getAudioChatMessageContentRef());
    }
  }

  private getChatMessageContentRef(): python.Reference {
    return python.reference({
      name:
        this.chatMessageContent.type + (this.isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });
  }

  private getStringChatMessageContentRef(): python.Reference {
    return python.reference({
      name: "StringChatMessageContent" + (this.isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });
  }

  private getFunctionCallChatMessageContentRef(): python.Reference {
    return python.reference({
      name:
        "FunctionCallChatMessageContent" +
        (this.isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });
  }

  private getFunctionCallChatMessageContentValueRef(): python.Reference {
    return python.reference({
      name:
        "FunctionCallChatMessageContentValue" +
        (this.isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });
  }

  private getAudioChatMessageContentRef(): python.Reference {
    return python.reference({
      name: "AudioChatMessageContent" + (this.isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });
  }

  public write(writer: Writer): void {
    const contentType = this.chatMessageContent.type;

    if (contentType === "STRING") {
      const stringContentValue = this.chatMessageContent.value;

      const stringChatMessageContentRequestRef =
        this.getStringChatMessageContentRef();
      stringChatMessageContentRequestRef.write(writer);
      writer.write(`(value="${stringContentValue}")`);
      return;
    }

    if (contentType === "FUNCTION_CALL") {
      const functionCallChatMessageContentValue = this.chatMessageContent.value;

      const functionCallChatMessageContentValueArgs: MethodArgument[] = [];

      if (functionCallChatMessageContentValue.id !== undefined) {
        functionCallChatMessageContentValueArgs.push(
          new MethodArgument({
            name: "id",
            value: python.TypeInstantiation.str(
              functionCallChatMessageContentValue.id
            ),
          })
        );
      }

      functionCallChatMessageContentValueArgs.push(
        new MethodArgument({
          name: "name",
          value: python.TypeInstantiation.str(
            functionCallChatMessageContentValue.name
          ),
        })
      );

      if (functionCallChatMessageContentValue.arguments !== undefined) {
        functionCallChatMessageContentValueArgs.push(
          new MethodArgument({
            name: "arguments",
            value: new Json(functionCallChatMessageContentValue.arguments),
          })
        );
      }

      const functionCallChatMessageContentValueRequestRef =
        this.getFunctionCallChatMessageContentValueRef();

      const functionCallChatMessageContentValueInstance =
        python.instantiateClass({
          classReference: functionCallChatMessageContentValueRequestRef,
          arguments_: functionCallChatMessageContentValueArgs,
        });

      const functionCallChatMessageContentRequestRef =
        this.getFunctionCallChatMessageContentRef();

      const functionCallChatMessageContentRequestInstance =
        python.instantiateClass({
          classReference: functionCallChatMessageContentRequestRef,
          arguments_: [
            new MethodArgument({
              name: "value",
              value: functionCallChatMessageContentValueInstance,
            }),
          ],
        });

      functionCallChatMessageContentRequestInstance.write(writer);
      return;
    }

    if (contentType === "ARRAY") {
      // TODO: Implement array types and call this recursively
      //    https://app.shortcut.com/vellum/story/4937/flesh-out-codegen-for-all-chat-message-content-types
      throw new Error("Unhandled type: ARRAY");
    }

    if (contentType === "IMAGE") {
      // TODO: Implement image types
      //    https://app.shortcut.com/vellum/story/4937/flesh-out-codegen-for-all-chat-message-content-types
      throw new Error("Unhandled type: IMAGE");
    }

    if (contentType === "AUDIO") {
      const audioContentValue = this.chatMessageContent.value;

      const audioChatMessageContentRequestRef =
        this.getAudioChatMessageContentRef();
      audioChatMessageContentRequestRef.write(writer);
      writer.write(`(value="${audioContentValue}")`);
      return;
    }

    assertUnreachable(contentType);
  }
}
