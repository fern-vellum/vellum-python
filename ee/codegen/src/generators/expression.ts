import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { NodeAttributeGenerationError } from "src/generators/errors";

export declare namespace Expression {
  interface Args {
    lhs: AstNode;
    expression: string;
    rhs?: AstNode | undefined;
    base?: AstNode | undefined;
  }
}

export class Expression extends AstNode {
  private readonly astNode: AstNode;

  constructor({ lhs, expression, rhs, base }: Expression.Args) {
    super();
    this.astNode = this.generateAstNode({ lhs, expression, rhs, base });
  }

  private generateAstNode({
    lhs,
    expression,
    rhs,
    base,
  }: Expression.Args): AstNode {
    this.inheritReferences(lhs);
    if (rhs) {
      this.inheritReferences(rhs);
    }

    // TODO: We should ideally perform this using native fern functionality, but it requires being able to create
    //  a Reference object from an existing AstNode, which in turn requires all AstNode's to internally track their
    //  name and modulePath.

    const rawExpression = base
      ? this.generateExpressionWithBase(base, expression, lhs, rhs)
      : this.generateStandardExpression(lhs, expression, rhs);

    return python.codeBlock(rawExpression);
  }

  private generateExpressionWithBase(
    base: AstNode,
    expression: string,
    lhs: AstNode,
    rhs: AstNode | undefined
  ): string {
    if (!rhs) {
      throw new NodeAttributeGenerationError(
        "rhs must be defined if base is defined"
      );
    }
    this.inheritReferences(base);
    return `${base.toString()}.${expression}(${lhs.toString()}, ${rhs.toString()})`;
  }

  private generateStandardExpression(
    lhs: AstNode,
    expression: string,
    rhs: AstNode | undefined
  ): string {
    const rhsExpression = rhs ? `(${rhs.toString()})` : "()";
    return `${lhs.toString()}.${expression}${rhsExpression}`;
  }

  public write(writer: Writer) {
    this.astNode.write(writer);
  }
}
