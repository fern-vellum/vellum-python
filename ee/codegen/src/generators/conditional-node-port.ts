import { python } from "@fern-api/python-ast";
import { MethodArgument } from "@fern-api/python-ast/MethodArgument";
import { OperatorType } from "@fern-api/python-ast/OperatorType";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";
import { isNil } from "lodash";

import { PortContext } from "src/context/port-context";
import { Expression } from "src/generators/expression";
import { NodeInput } from "src/generators/node-inputs";
import {
  ConditionalNodeConditionData,
  ConditionalRuleData,
} from "src/types/vellum";

export declare namespace ConditionalNodePort {
  export interface Args {
    portContext: PortContext;
    inputFieldKeysByRuleId: Map<string, string>;
    valueInputKeysByRuleId: Map<string, string>;
    conditionData: ConditionalNodeConditionData;
    nodeInputsByKey: Map<string, NodeInput>;
  }
}

export class ConditionalNodePort extends AstNode {
  private portContext: PortContext;
  private conditionalNodeData: ConditionalNodeConditionData;
  private inputFieldKeysByRuleId: Map<string, string>;
  private valueInputKeysByRuleId: Map<string, string>;
  private nodeInputsByKey: Map<string, NodeInput>;
  private astNode: AstNode;

  public constructor(args: ConditionalNodePort.Args) {
    super();

    this.portContext = args.portContext;
    this.inputFieldKeysByRuleId = args.inputFieldKeysByRuleId;
    this.valueInputKeysByRuleId = args.valueInputKeysByRuleId;
    this.conditionalNodeData = args.conditionData;
    this.nodeInputsByKey = args.nodeInputsByKey;
    this.astNode = this.constructPort();
    this.inheritReferences(this.astNode);
  }

  private constructPort(): AstNode {
    return python.invokeMethod({
      methodReference: python.reference({
        name: "Port",
        modulePath:
          this.portContext.workflowContext.sdkModulePathNames.PORTS_MODULE_PATH,
        attribute: [
          this.convertConditionTypeToPortAttribute(
            this.conditionalNodeData.type
          ),
        ],
      }),
      arguments_: (() => {
        const arg = this.generatePortCondition();
        return arg ? [arg] : [];
      })(),
    });
  }

  private generatePortCondition() {
    const conditionData = this.conditionalNodeData.data;
    if (conditionData) {
      return new MethodArgument({
        value: this.buildCondition(conditionData),
      });
    } else {
      return undefined;
    }
  }

  private buildCondition(
    conditionData: ConditionalRuleData | undefined
  ): AstNode {
    if (!conditionData) {
      return python.TypeInstantiation.none();
    }

    if (conditionData && conditionData.fieldNodeInputId) {
      return this.buildDescriptor(conditionData);
    }

    const otherConditions = (conditionData.rules || []).map((rule) => {
      return this.buildCondition(rule);
    });

    const combine =
      conditionData.combinator === "AND"
        ? (lhs: AstNode, rhs: AstNode): AstNode => {
            return python.operator({
              operator: OperatorType.And,
              lhs: lhs,
              rhs: rhs,
            });
          }
        : (lhs: AstNode, rhs: AstNode): AstNode => {
            return python.operator({
              operator: OperatorType.Or,
              lhs: lhs,
              rhs: rhs,
            });
          };

    const combinedConditions = otherConditions.reduce((prev, curr) => {
      return combine(prev, curr);
    });
    return combinedConditions;
  }

  private convertConditionTypeToPortAttribute(conditionType: string): string {
    switch (conditionType) {
      case "IF":
        return "on_if";
      case "ELIF":
        return "on_elif";
      default:
        return "on_else";
    }
  }

  private buildDescriptor(conditionData: ConditionalRuleData): AstNode {
    const ruleId = conditionData.id;

    const lhsKey = this.inputFieldKeysByRuleId.get(ruleId);
    let rhsKey;
    if (isNil(lhsKey)) {
      throw new Error(`Could not find input field key given rule id ${ruleId}`);
    }
    if (conditionData.valueNodeInputId) {
      rhsKey = this.valueInputKeysByRuleId.get(ruleId);
    }
    const lhs = this.nodeInputsByKey.get(lhsKey);
    const rhs = !isNil(rhsKey) ? this.nodeInputsByKey.get(rhsKey) : undefined;
    const expression = conditionData.operator
      ? this.convertOperatorToMethod(conditionData.operator)
      : undefined;
    if (isNil(lhs) || isNil(expression)) {
      throw new Error("Port conditions require a lhs and an expression");
    }
    return new Expression({
      lhs: lhs,
      expression: expression,
      rhs: rhs,
    });
  }

  private convertOperatorToMethod(operator: string): string {
    const operatorMappings: { [key: string]: string } = {
      "=": "equals",
      "!=": "does_not_equal",
      "<": "less_than",
      ">": "greater_than",
      "<=": "less_than_or_equal_to",
      ">=": "greater_than_or_equal_to",
      contains: "contains",
      beginsWith: "begins_with",
      endsWith: "ends_with",
      doesNotContain: "does_not_contain",
      doesNotBeginWith: "does_not_begin_with",
      doesNotEndWith: "does_not_end_with",
      null: "is_none",
      notNull: "is_not_none",
      in: "in",
      notIn: "not_in",
      between: "between",
      notBetween: "not_between",
    };
    const value = operatorMappings[operator];
    if (!value) {
      throw new Error(`This operator: ${operator} is not supported`);
    }
    return value;
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}
