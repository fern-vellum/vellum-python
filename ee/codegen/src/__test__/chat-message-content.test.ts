import { Writer } from "@fern-api/python-ast/core/Writer";

import { ChatMessageContent } from "src/generators/chat-message-content";

describe("ChatMessageContent", () => {
  let writer: Writer;

  beforeEach(() => {
    writer = new Writer();
  });

  describe("STRING", () => {
    it("should write a string content correctly", async () => {
      const chatMessageContent = new ChatMessageContent({
        chatMessageContent: { type: "STRING", value: "Hello, AI!" },
      });
      chatMessageContent.write(writer);
      expect(await writer.toString()).toMatchSnapshot();
    });
  });

  describe("FUNCTION_CALL", () => {
    it("should write a function call content with id correctly", async () => {
      const chatMessageContent = new ChatMessageContent({
        chatMessageContent: {
          type: "FUNCTION_CALL",
          value: {
            id: "123",
            name: "get_weather",
            arguments: {
              location: "New York",
              unit: "celsius",
            },
          },
        },
      });
      chatMessageContent.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should write a function call content without id correctly", async () => {
      const chatMessageContent = new ChatMessageContent({
        chatMessageContent: {
          type: "FUNCTION_CALL",
          value: {
            name: "get_weather",
            arguments: {
              location: "New York",
              unit: "celsius",
            },
          },
        },
      });
      chatMessageContent.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("AUDIO", () => {
    it("should write an audio content correctly", async () => {
      const chatMessageContent = new ChatMessageContent({
        chatMessageContent: {
          type: "AUDIO",
          value: {
            src: "https://example.com/audio.mp3",
            metadata: { key: "value" },
          },
        },
      });
      chatMessageContent.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
