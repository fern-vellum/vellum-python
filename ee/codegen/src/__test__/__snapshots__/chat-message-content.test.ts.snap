// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatMessageContent > AUDIO > should write an audio content correctly 1`] = `
"AudioChatMessageContentRequest(
    src="https://example.com/audio.mp3", metadata={"key": "value"}
)
"
`;

exports[`ChatMessageContent > FUNCTION_CALL > should write a function call content with id correctly 1`] = `
"FunctionCallChatMessageContentRequest(
    value=FunctionCallChatMessageContentValueRequest(
        id="123",
        name="get_weather",
        arguments={"location": "New York", "unit": "celsius"},
    )
)
"
`;

exports[`ChatMessageContent > FUNCTION_CALL > should write a function call content without id correctly 1`] = `
"FunctionCallChatMessageContentRequest(
    value=FunctionCallChatMessageContentValueRequest(
        name="get_weather", arguments={"location": "New York", "unit": "celsius"}
    )
)
"
`;

exports[`ChatMessageContent > STRING > should write a string content correctly 1`] = `"StringChatMessageContentRequest(value="Hello, AI!")"`;
